src/DxbcBuilder.cpp

```diff
@@ -0,0 +1,197 @@
+}
+
+//---------------------------------------------------------------------------------------------------------------------------------
+// CDXBCBuilder::AppendBlob
+HRESULT CDXBCBuilder::AppendBlob(CDXBCParser *pParser, DXBCFourCC BlobFourCC)
+{
+    if (!pParser)
+        return E_FAIL;
+
+    UINT blobIndex = pParser->FindNextMatchingBlob(BlobFourCC);
+    if (blobIndex == DXBC_BLOB_NOT_FOUND)
+        return S_FALSE;
+
+    const void *pData = pParser->GetBlob(blobIndex);
+    UINT cbData = pParser->GetBlobSize(blobIndex);
+
+    if (!pData)
+        return E_FAIL;
+
+    return AppendBlob(BlobFourCC, cbData, pData);
+}
+
+//---------------------------------------------------------------------------------------------------------------------------------
+// CDXBCBuilder::GetFinalDXBC
+HRESULT CDXBCBuilder::GetFinalDXBC(void *pCallerAllocatedMemory, UINT *pContainerSize)
+{
+    if (!pCallerAllocatedMemory)
+    {
+        // Return how much memory the caller needs to allocate.
+        if (pContainerSize)
+        {
+            *pContainerSize = m_TotalOutputContainerSize;
+            return S_OK;
+        }
+        return E_FAIL;
+    }
+    if (!pContainerSize)
+    {
+        return E_FAIL; // Nothing to do.
+    }
+    if (*pContainerSize < m_TotalOutputContainerSize)
+    {
+        // not enough memory allocated, return 0 bytes written out.
+        *pContainerSize = 0;
+        return E_FAIL;
+    }
+    // Ok, we can write out the full container.
+    DXBCHeader *pHeader = (DXBCHeader *)pCallerAllocatedMemory;
+    UINT *pIndex = (UINT *)((BYTE *)pHeader + sizeof(DXBCHeader));                                     // skip past header
+    DXBCBlobHeader *pNextBlobHeader = (DXBCBlobHeader *)((BYTE *)pIndex + m_BlobCount * sizeof(UINT)); // skip past index
+    BYTE *pNextBlobData = (BYTE *)pNextBlobHeader + sizeof(DXBCBlobHeader);                            // skip past blob header
+    BlobNode *pNextInputBlob = m_pFirstBlob;
+    for (UINT b = 0; b < m_BlobCount; b++)
+    {
+        pIndex[b] = (UINT)((BYTE *)pNextBlobHeader - (BYTE *)pCallerAllocatedMemory);
+        *pNextBlobHeader = pNextInputBlob->BlobHeader;
+        memcpy(pNextBlobData, pNextInputBlob->pBlobData, pNextInputBlob->BlobHeader.BlobSize);
+
+        // advance to next blob
+        pNextBlobHeader = (DXBCBlobHeader *)(pNextBlobData + pNextInputBlob->BlobHeader.BlobSize); // skip past last blob's data
+        pNextBlobData = (BYTE *)pNextBlobHeader + sizeof(DXBCBlobHeader);                          // skip past blob header
+        pNextInputBlob = pNextInputBlob->pNext;
+    }
+
+    // Fill in the initial entries
+    pHeader->BlobCount = m_BlobCount;
+    pHeader->ContainerSizeInBytes = m_TotalOutputContainerSize;
+    pHeader->DXBCHeaderFourCC = DXBC_FOURCC_NAME;
+    pHeader->Version.Major = DXBC_MAJOR_VERSION;
+    pHeader->Version.Minor = DXBC_MINOR_VERSION;
+
+    if (FAILED(SignDxbc((BYTE *)pCallerAllocatedMemory, pHeader->ContainerSizeInBytes)))

```

Comment:
I think we probably want to make it so that signing is a post-processing step on building, so the signer component only needs to be included if you need signed DXBC.

Reply:
Ah that makes sense, I like that approach more too.